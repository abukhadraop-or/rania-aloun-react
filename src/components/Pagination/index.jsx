import React from 'react';
import _ from 'lodash';
import ProbTypes from 'prop-types';
import config from 'config';
import {
  PagesList,
  PageNumber,
  PageButton,
  PagesContainer,
} from './pagination.styles';

/**
 * Pagination component for showing number of pages at the bottom of the page.
 *
 * @param {Object}   props Props passed to the pagination component.
 * @param {number}   props.totalArticles Number of articles to be displayed and paginated.
 * @param {Function} props.onPageChange Changes the selected page number.
 *
 * @return {JSX.Element} Pages numbers in a component at the bottom of the page.
 */
function Pagination({ totalArticles, onPageChange }) {
  let isNeeded = true;

  /**
   * Calculates the needed page numbers, if there is only one page, it toggles the isNeeded state to hide the pagination component.
   *
   * @return {Number[]} Array holding pages numbers to be rendered.
   */
  const paginate = () => {
    const pagesCount = Math.ceil(totalArticles / config.pageSize);
    if (pagesCount === 1) {
      isNeeded = false;
    }
    const pages = _.range(0, pagesCount);
    return pages;
  };

  const pages = paginate();

  return (
    isNeeded && (
      <PagesContainer>
        <PagesList>
          {pages.map((page) => (
            <PageNumber key={page}>
              <PageButton onClick={() => onPageChange(page)}>{page}</PageButton>
            </PageNumber>
          ))}
        </PagesList>
      </PagesContainer>
    )
  );
}

Pagination.propTypes = {
  totalArticles: ProbTypes.number.isRequired,
  onPageChange: ProbTypes.func.isRequired,
};

export default Pagination;
