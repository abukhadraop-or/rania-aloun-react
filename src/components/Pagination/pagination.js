import styled from "styled-components";

export const PagesContainer = styled.nav`
  margin: 2rem 0 0;
`;

export const PagesList = styled.ul`
  display: flex;
  flex-wrap: wrap;
  list-style-type: none;
  margin: 0;
  padding: 0;
`;

export const PageNumber = styled.li`
  border-radius: 0;
  border: 0.063rem solid #bbbbbb;
  color: #5cb85c;
  cursor: pointer;
  margin: 0 -0.063rem 0 0;
  width: fit-content;
  &:active {
    color: white;
    background-color: #5cb85c;
    border: none;
  }
`;

export const PageButton = styled.button`
  background-color: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  font-size: 1rem;
  padding: 0.6rem;
`;
