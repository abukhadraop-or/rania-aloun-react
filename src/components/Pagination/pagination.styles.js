import styled from 'styled-components';
import colors from 'styles/colors';

export const PagesContainer = styled.nav`
  margin: 2rem 0 0;
`;

export const PagesList = styled.ul`
  display: flex;
  flex-wrap: wrap;
  list-style-type: none;
  margin: 0;
  padding: 0;
`;

export const PageNumber = styled.li`
  border-radius: 0;
  border: 0.063rem solid ${colors.ashGrey};
  color: ${colors.boringGreen};
  cursor: pointer;
  margin: 0 -0.063rem 0 0;
  width: fit-content;
  &:active {
    color: white;
    background-color: ${colors.boringGreen};
    border: none;
  }
`;

export const PageButton = styled.button`
  background-color: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  font-size: 1rem;
  padding: 0.6rem;
`;
