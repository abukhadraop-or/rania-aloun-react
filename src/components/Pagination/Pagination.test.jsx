import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Pagination from '.';

describe('<Pagination />', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(
      <Pagination totalArticles={8} onPageChange={jest.fn()} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
