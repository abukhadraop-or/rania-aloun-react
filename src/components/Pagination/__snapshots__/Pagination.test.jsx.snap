// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> matches snapshot 1`] = `
<styled.nav>
  <styled.ul>
    <styled.li
      key="0"
    >
      <styled.button
        onClick={[Function]}
      >
        0
      </styled.button>
    </styled.li>
    <styled.li
      key="1"
    >
      <styled.button
        onClick={[Function]}
      >
        1
      </styled.button>
    </styled.li>
    <styled.li
      key="2"
    >
      <styled.button
        onClick={[Function]}
      >
        2
      </styled.button>
    </styled.li>
    <styled.li
      key="3"
    >
      <styled.button
        onClick={[Function]}
      >
        3
      </styled.button>
    </styled.li>
  </styled.ul>
</styled.nav>
`;
